### PR TITLE
[Perf] Avoid forward scan for async output placeholders

### PR DESCRIPTION
**Pull request by izikgo**
## Purpose

This PR avoids a forward scan through each request's full generated-token
history when patching async output-token placeholders.
`InputBatch.update_async_output_token_ids()` currently finds the first `-1`
placeholder with `req_output_token_ids.index(-1)`, which is linear in the
generated sequence length for each running request.

Because async output-token placeholders are appended at the end of each
request's output history, this patch scans backward over the trailing
placeholder run instead. That makes the work proportional to the number of
placeholders, which is typically one for regular decode and bounded by the
speculative token count for spec decode.

For long-generation workloads, the old implementation can add a large Python
CPU gap every decode step: `O(num_reqs * output_len)`. This is most visible
when async scheduling is enabled and the request configuration requires
generated-token history to be available during sampling, for example reasoning
parsers, custom logits processors, penalties, or bad-words filtering.

In V1, `SamplingMetadata.output_token_ids` is populated by
`InputBatch._make_sampling_metadata()` when any of these conditions is true:
penalties are enabled (`not self.no_penalties`), bad-word token IDs are present,
logits processors may need output token IDs, or the thinking-budget state holder
has tracked requests. The logits-processor condition is set conservatively by
`GpuModelRunner`: it is true when custom logits processors are configured, or
when `vllm_config.reasoning_config is not None`. The server-level
`--reasoning-parser` flag creates that `ReasoningConfig`, so a reasoning parser
is enough to keep `output_token_ids` in sampling metadata even for
`/v1/completions` requests.

The new path preserves the existing behavior while avoiding the long-history
scan.

This PR was prepared with AI assistance and manually reviewed.

## Test Plan

Syntax check:

```bash
python3 -m py_compile vllm/v1/worker/gpu_input_batch.py
```

Manual synthetic timing and an end-to-end serving benchmark are included below.

## Test Results

Syntax check passed.

Manual timing of the same synthetic case showed the old forward-scan shape at
about 10.9 ms and the fixed path at about 0.14 ms.

I also checked the behavior with a local vLLM server load that exercises the
actual async output-token-history path. The setup used `vllm serve` with async
scheduling and `vllm bench serve --backend vllm` with 128 concurrent completion
requests, 4096 input tokens, 60000 requested output tokens, and
`--reasoning-parser nemotron_v3`. The reasoning parser is sufficient to keep
`output_token_ids` in sampling metadata, so `update_async_output_token_ids()`
runs every decode step.

This run intentionally used 60000 output tokens rather than a larger 120k-131k
case so that the fast/fixed run would not run into KV-cache pressure before
completion.

Server:

```bash
vllm serve nvidia/NVIDIA-Nemotron-3-Nano-30B-A3B-BF16 \
  --served-model-name cpu-list-bench \
  --load-format dummy \
  --max-model-len 131072 \
  --max-num-seqs 128 \
  --max-num-batched-tokens 131072 \
  --gpu-memory-utilization 0.8 \
  --async-scheduling \
  --reasoning-parser nemotron_v3 \
  --trust-remote-code \
  --generation-config vllm \
  --no-enable-log-requests \
  --no-enable-chunked-prefill \
  --no-enable-prefix-caching
```

Client:

```bash
vllm bench serve \
  --backend vllm \
  --model nvidia/NVIDIA-Nemotron-3-Nano-30B-A3B-BF16 \
  --served-model-name cpu-list-bench \
  --dataset-name random \
  --random-input-len 4096 \
  --random-output-len 60000 \
  --num-prompts 128 \
  --request-rate inf \
  --max-concurrency 128 \
  --ignore-eos \
  --trust-remote-code \
  --temperature 0
```

Completed benchmark results:

| Image | Completed | Output tokens | Duration | Output throughput | Mean TPOT | P99 ITL |
| --- | ---: | ---: | ---: | ---: | ---: | ---: |
| v0.20.0 clean | 128/128 | 7,680,000 | 3965.3 s | 1936.8 tok/s | 65.48 ms | 197.32 ms |
| v0.20.0 + this patch | 128/128 | 7,680,000 | 630.6 s | 12178.1 tok/s | 9.89 ms | 13.08 ms |

Both runs generated the same number of output tokens with no failed requests.
The server logs reported `Waiting: 0 reqs` throughout both runs and max GPU KV
cache usage of 79.2% for both runs. The clean image steadily slowed as output
histories grew, ending around 0.8k output tok/s in the server log. The patched
image avoided the much steeper CPU-list scan degradation and was still around
10.2k output tok/s shortly before completion.

### CODE REVIEW

File: tests/v1/worker/test_gpu_input_batch_perf.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+from types import SimpleNamespace
+
+import pytest
+import torch
+
+from vllm.v1.worker.gpu_input_batch import InputBatch
+
+
+@pytest.mark.cpu_test
+def test_update_async_output_token_ids_handles_long_outputs_quickly() -> None:
```
Comment (MatthewBonanni) [inline tests/v1/worker/test_gpu_input_batch_perf.py]: nit: this test probably isn't necessary and will likely be flaky due to hardcoded 5ms limit. I think we can just remove it

**Review by MatthewBonanni — COMMENTED** [review]
LGTM once comment is addressed, thanks for finding this! I am skeptical of the e2e benchmark numbers though -- this would imply that this one list scan was 84% of total wall time. What hardware did you run this on?

Definitely a nice neat optimization, worth merging regardless

**Comment by izikgo** [comment]
@MatthewBonanni Thank you for your feedback.
I ran it on a single B200.
I don't think the list traversal itself it what eats the wall time. The large delta is because in async mode the CPU work is normally masked by GPU forward time. Once the list scan grows long enough to exceed the GPU step time, it lands on the critical path and creates gaps between forwards, so throughput falls off sharply.

File: tests/v1/worker/test_gpu_input_batch_perf.py (hunk 1 above)
Comment (izikgo) [reply tests/v1/worker/test_gpu_input_batch_perf.py]: removed

**Review by MatthewBonanni — APPROVED** [review]
LGTM, thanks!
